proposals/ref-assignment-switch-expressions.md

```diff
@@ -0,0 +1,69 @@
+# `ref` assignment on `switch` expressions
+
+* [x] Proposed
+* [ ] Prototype: None
+* [ ] Implementation: None
+* [ ] Specification: Started, below
+
+## Summary
+[summary]: #summary
+
+Provide the ability to return `ref` values in `switch` expressions.
+
+## Motivation
+[motivation]: #motivation
+
+The `switch` expression is meant to provide a quicker way to return values based on a given expression, but does not support returning `ref` expressions. This limitation doesn't prevent any harm, and can be lifted.
+
+## Detailed design
+[design]: #detailed-design
+
+A `switch` expression may return a `ref` value. The returning expression, if not a `throw` expression, must be a `ref` expression.
+
+A `switch` expression that returns a `ref` value needs an extra `ref` in front of it when assigned/returned to a `ref` local. The examples below cover this case too. This design aligns with the current design in the ternary operator.
+
+`throw` expressions can still be used normally.
+
+## Examples
+[examples]: #examples
+
+```csharp
+private ref int GetDirectionField(Direction direction) => ref direction switch
+{
+     Direction.North => ref NorthField,
+     Direction.South => ref SouthField,
+     Direction.East => ref EastField,
+     Direction.West => ref WestField,
+     _ => throw new NotImplementedException(),
+};
+
+private void RefSwitchAssignLocal()
+{
+    ref int dimension = ref axis switch
+    {
+        Axis.X => ref X,
+        Axis.Y => ref Y,
+    };
+}
```

Comment:
This design needs to cover the lifetime implications as it deals with `ref` semantics. Essentially how do we calculate the *safe-to-escape* of the return here?

Reply:
I don't get that concern, is it relevant to ref scoped? I'm out of the loop with that feature.

Reply:
I would expect all the ref-safety caveats of ref-`?:` to apply here. Including that RSTE/STE is equal to the narrowest of all switch arms's RSTE/STE, and that when the ref is mutable, the STEs of the switch arms must be equal.

Reply:
thanks @RikkiGibson !

Reply:
That still didn't clear it out to me, what exactly are you referring to with RSTE/STE? I can only infer about the level of access on the referenced memory.

Reply:
STE is Safe-To-Escape and RSTE is Ref-Safe-To-Escape.

Reply:
Included details about RSTE/STE